Add backend for global CE Admin

martha opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

frontend pr https://github.com/greenriver/hmis-frontend/pull/1151

## Description
- Opportunity:
  - Add new fields including organization name, unit, and date available
  - Add new sort options: ascending/descending based on date available. (For now this just reflects the date the opportunity was generated)
  - Add filter options including organization, available date, and workflow template
- Referral:
  - Add new fields, including resolving `current_steps` plural instead of `current_step`
  - Add papertrail to support `updated_by`. Also add papertrail to Steps
  - Add filter options including organization, template, and length of time it's been on the current step
- Factor out common code for resolving filters, etc. into concerns `HasCeReferrals` and `HasCeOpportunities`

## Type of change
New feature

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] If adding a new endpoint / exposing data in a new way, I have:
  - [x] ensured the API can't leak data from other data sources
  - [x] ensured this does not introduce N+1s
  - [x] ensured permissions and visibility checks are performed in the right places
- [x] Any major architectural changes are supported by an approved ADR (Architectural Decision Record) 
- [x] I have updated the documentation (or not applicable)
- [x] I have added spec tests (or not applicable)
- [x] I have provided testing instructions in this PR or the related issue (or not applicable)

[//]: # NOTE: system tests may fail if there is no branch on the hmis-frontend that matches the Source  or Target branch of this PR. This is expected

martha committed:
Add Admin Opportunities

martha committed:
Add Admin Referrals (wip)

martha committed:
Improve filters and tests

martha committed:
Remove template filter test

martha committed:
Solve template filter

martha commented on drivers/hmis/app/graphql/types/hmis_schema/has_ce_opportunities.rb:
```diff
@@ -0,0 +1,52 @@
+
+        scope = scope.viewable_by(user)
+
+        scope = scope.where(status: filters&.status) if filters.respond_to?(:status) && filters&.status.present?
```
To discuss: I don't really love this messy `filter.respond_to?` check peppered in everywhere. I was wondering if we have a common pattern for dealing with this; maybe that's what the `yield_self` in `EnrollmentFilter` helps with? https://github.com/greenriver/hmis-warehouse/blob/stable/drivers/hmis/app/models/hmis/filter/enrollment_filter.rb

gigxz replied:
I think that's what the `with_filter` helper does in the filter classes (like the one you linked). It might be good to follow the pattern of having a dedicated class for filtering, since this has a lot of filter capabilities. It lets us test filters in isolation without graphql.

martha replied:
Thanks! Hm do you have any examples in the code you can link me to where we are testing filters in isolation without graphql? I wasn't finding any instances of `apply_filters` in the spec files.

gigxz replied:
Ha I guess just because we _can_ in theory  doesn't mean we are 👀

martha replied:
Ok sounds good haha! For now I'll leave this as-is with tests in the request spec files.

martha commented on drivers/hmis/app/graphql/types/hmis_schema/ce_referral.rb:
```diff
@@ -57,20 +64,24 @@ def client
 
-      step = load_ar_association(instance, :steps, scope: step_scope).first
-      return if step.nil?
+    def updated_by
```
This is just a sketch of an approach we might take here (but somewhat outside of spec). Let me know what you think!

gigxz replied:
Thanks. I'm not sure if there is a meaningful distinction between the "updated_at" on ce_referrals vs the "updated_at"  on each step. If not, we could take the approach of updating `updated_at` on the referral table any time on of the sub tasks is touched? I also think we will want to add `last_updated_by_user_id` references (and likely created_by), because we will want to use those for reporting. That definitely feels out of scope here, though. I think you can leave it as-is or stubbed out, and add a reference to https://github.com/open-path/Green-River/issues/7678

martha commented on drivers/hmis/app/graphql/types/hmis_schema/ce_opportunity.rb:
```diff
@@ -53,6 +59,21 @@ def project_type
+    end
+
+    def date_available
+      # TODO(#7537) - implement "available after date". Always returns date the referral was created, for now
```
My understanding, from the spec's reference to “recency”, is that this is meant to be an amalgamation of the opportunity's creation date with the (not yet implemented) date available field. So I've implemented it as `created_at` for now with todos for the future ticket.

martha commented on drivers/hmis/app/graphql/types/forms/pick_list_option.rb:
```diff
@@ -162,12 +162,21 @@ def self.static_options_for_type(pick_list_type, user:)
         Hmis::WorkflowDefinition::Template.all.map do |template|
           { code: template.id, label: template.name }
         end
+      when 'WORKFLOW_DEFINITION_TEMPLATE_IDENTIFIERS'
```
Happy to discuss/change this around; for example, we could replace the original `WORKFLOW_DEFINITION_TEMPLATES` with the new one, instead of adding `WORKFLOW_DEFINITION_TEMPLATE_IDENTIFIERS`. (Related to considerations about when to remove [this](https://github.com/greenriver/hmis-warehouse/pull/5336))

gigxz replied:
It makes sense to me that the dropdown used for filtering would be different from the dropdown used for configuration, since the filtering one includes retired templates but the configuration one should not. However we should probably indicate in the pick list name that it includes retired templates.

I would expect we end up with something like:
* `CE_WORKFLOW_TEMPLATE_IDENTIFIERS_INCLUDING_RETIRED` - unique ce workflow template identifiers, including representing workflows that are completly retired. used for filtering on existing/historical referrals..
* `CE_WORKFLOW_TEMPLATE_IDENTIFIERS` - unique ce workflow template identifiers that are currently published. used for configuring which template to use for a project/resource/group/etc (tbd https://github.com/open-path/Green-River/issues/7522)

martha commented on drivers/hmis/lib/tasks/ce_starter_pack_20250302.rake:
```diff
@@ -201,6 +201,7 @@ task ce_starter_pack_20250302: [:environment] do
   start_workflow_event = create_start_event(enrollment_creator_template, 'start referral', 'start_workflow', 'start_referral')
   accept_workflow_event = create_end_event(enrollment_creator_template, 'accept referral', 'end_workflow', 'accept_referral')
 
+  client_acceptance_task = create_task(client_accepts_form_def, enrollment_creator_template,  'Confirm Client Accepts Referral', case_managers)
```
These changes to the starter pack update the Enrollment Creator template to have multiple steps in progress at the same time, just to show what that looks like. 
![Screenshot 2025-05-06 at 6 27 35 PM](https://github.com/user-attachments/assets/04f683f1-9df4-49eb-bafc-2f4a8adf3a96)

martha committed:
PR cleanup

martha committed:
Merge branch 'release-163' into mke/7612-ce-admin

martha committed:
Fix bugs

martha committed:
Merge branch 'release-163' into mke/7612-ce-admin

gigxz commented on drivers/hmis/app/graphql/types/forms/pick_list_option.rb:
```diff
@@ -162,12 +162,21 @@ def self.static_options_for_type(pick_list_type, user:)
+      when 'WORKFLOW_DEFINITION_TEMPLATE_IDENTIFIERS'
+        # used by dropdowns for filters when filtering referral/opportunity by workflow template.
+        # Here we want to group templates by identifier in case there are multiple/past versions.
+        Hmis::WorkflowDefinition::Template.published.or(Hmis::WorkflowDefinition::Template.retired).
```
These should filter on `template_type: :ce_referral`

gigxz commented on drivers/hmis/app/graphql/types/forms/pick_list_option.rb:
```diff
@@ -162,12 +162,21 @@ def self.static_options_for_type(pick_list_type, user:)
+        # Here we want to group templates by identifier in case there are multiple/past versions.
+        Hmis::WorkflowDefinition::Template.published.or(Hmis::WorkflowDefinition::Template.retired).
+          group_by(&:identifier).map do |identifier, templates|
+          description = templates.max_by(&:updated_at).name
```
We could alternatively just pick the published one, right? I think its a little confusing to use updated_at as a proxy for that. (Assuming we have the right validations in place to ensure exactly 1 published template per identifier)

martha replied:
Sure sounds good, but we will need a fallback since there may not be a published version. It's probably more suitable to fallback to max by version number, instead of updated timestamp.

gigxz commented on drivers/hmis/app/graphql/types/hmis_schema/ce_opportunity.rb:
```diff
@@ -19,17 +19,23 @@ class HmisSchema::CeOpportunity < Types::BaseObject
     field :categories, [String], null: false
     field :active, Boolean, null: false, method: :active?
     field :candidates_generated_at, GraphQL::Types::ISO8601DateTime, null: true
+    field :date_available, GraphQL::Types::ISO8601Date, null: true
```
Does this need to be nullable? All opportunities should have a created_at value

gigxz commented on drivers/hmis/app/graphql/types/hmis_schema/ce_referral.rb:
```diff
@@ -16,14 +16,18 @@ class HmisSchema::CeReferral < Types::BaseObject
-    field :target_enrollment, Types::HmisSchema::Enrollment, null: true # Don't resolve in batch
+
+    field :current_steps, [HmisSchema::CeReferralStep], null: true
+    field :current_step_time, GraphQL::Types::ISO8601DateTime, null: true
```
I think this field name could be a little more descriptive. It's not clear to me if it's the time that the step was started, or time that it was last updated, or time it was completed, etc. 

For populating a column "Days in Current Step(s)", which it looks like this is doing in the frontend,  I think it would be simpler to just do that calculation on the backend and resolve it as a field:

```ruby
# how many days has this referral been on the current step. If there are multiple open steps, it uses the oldest one
 def days_on_open_steps
    oldest_open_step = load_ar_association(object, :current_steps).to_a.min_by(&:created_at)
    # how many days ago was this step created
    opened_days_ago = (Date.current - oldest_open_step.created_at.to_date).to_i
    opened_days_ago
end
```

What do you think?

martha replied:
Cool, I agree, I'll make the update to resolve this in the backend and return it in the schema as an int number of days.

I'm not sure about using `created_at`, however. Seems like that could lead to inaccuracies when the steps are not made available at the same time as they are created. For example, in the circular workflow case, a step can get re-marked as available.

`updated_at` is not a perfect fit either, but fixing this seemed out of the current ticket's scope, so I created a placeholder over [here](https://github.com/open-path/Green-River/issues/7647) to make sure we don't forget to return to this. If it's ok with you, I will go with your approach but keep using `updated_at` instead of `created_at` for now.

gigxz replied:
I see, sounds like we need a new timestamp to capture the time when the step became available. Thanks for making a ticket

martha committed:
Move referral filters into CeReferralFilter

martha committed:
Update workflow definition template picklist

martha committed:
Make date available non-nullable

martha committed:
Resolve days on current step in backend

martha committed:
Add a filter class for opportunity too

martha committed:
Merge branch 'release-163' into mke/7612-ce-admin

martha committed:
Fix typo

gigxz commented on drivers/hmis/app/models/hmis/filter/ce_referral_filter.rb:
```diff
@@ -0,0 +1,52 @@
+
+  def on_current_step_since(scope)
+    with_filter(scope, :on_current_step_since) do
+      scope.joins(:current_steps).where(wfe_step_t[:updated_at].lt(input.on_current_step_since)).distinct
```
Add a TODO for https://github.com/open-path/Green-River/issues/7647 here?

gigxz commented on drivers/hmis/app/models/hmis/workflow_execution/instance.rb:
```diff
@@ -6,6 +6,7 @@ module Hmis::WorkflowExecution
   class Instance < GrdaWarehouseBase
     belongs_to :template, class_name: 'Hmis::WorkflowDefinition::Template'
     has_many :steps, class_name: 'Hmis::WorkflowExecution::Step', dependent: :destroy
+    has_many :open_steps, -> { open.order_by_updated_at }, class_name: 'Hmis::WorkflowExecution::Step'
```
Does this also need a TODO for 7647 to sort by the time the step was made available? It would be great if you can add a comment explaining why the sorting (ie if theres a particular schema field that depends on it)

gigxz replied:
Whoops nevermind I see you have the todo on the scope itself

martha replied:
There's no longer a schema field that grabs just the first one - I removed that in favor of returning them all and displaying them in a truncated list. I did add a comment though to try to clarify the purpose, thanks!

gigxz commented on drivers/hmis/spec/requests/hmis/ce/admin_ce_opportunities_spec.rb:
```diff
@@ -0,0 +1,76 @@
+      end
+    end
+
+    context 'with many opportunities' do
```
This is more of a "with many projects" test, since it is setting up 40 projects each with 1 opportunity (Or maybe the test could be expanded to have multiple opportunities per project?)

gigxz commented on drivers/hmis/app/graphql/types/hmis_schema/query_type.rb:
```diff
@@ -554,5 +556,17 @@ def ce_referral_step(id:)
+    ce_opportunities_field
+    def ce_opportunities(**args)
+      # TODO(#7506) - gated by admin permission
+      resolve_ce_opportunities(Hmis::Ce::Opportunity.all, **args)
```
Oh didn't think to call this out in the spec, but this needs to be limited to Opportunities that belong to Projects in the current Data Source. Per https://github.com/open-path/Green-River/issues/6612 we never want to be resolving data from other data sources

martha replied:
Hm thanks good point. Could we rely on the `viewable_by` scopes to do this restriction? They are not fully implemented yet, but I stubbed them out simply here in a way that I think achieves what we are after: https://github.com/greenriver/hmis-warehouse/pull/5360/commits/f2c91b075bcea4125c0f636d00154e4fc7b1e3f6

gigxz replied:
Sorry I lost track of this @martha, your implementation looks good to me.

gigxz commented on drivers/hmis/app/graphql/types/hmis_schema/query_type.rb:
```diff
@@ -554,5 +556,17 @@ def ce_referral_step(id:)
+
+    ce_referrals_field
+    def ce_referrals(**args)
+      resolve_ce_referrals(Hmis::Ce::Referral.all, **args)
```
Same here needs to be limited by current data source. thank you!

martha committed:
PR feedback improvements

martha committed:
Merge branch 'release-163' into mke/7612-ce-admin

martha committed:
Fix tests

martha committed:
Merge branch 'release-163' into mke/7612-ce-admin

martha commented:
@gigxz in the spirit of speedier development, I'd like to merge this now to the release branch, in order to resolve conflicts with the permissions work and get that in a good state to PR (and avoid double-dealing with conflicts, since we squash merge).

I think I've addressed all your concerns on this PR. For posterity, the outstanding question is about whether we can use `viewable_by` scope to limit queries by data source. This is also related to the permissions work, so we can continue the discussion on that PR if needed.

I plan to leave the frontend PR open for now https://github.com/greenriver/hmis-frontend/pull/1151#discussion_r2087082323 until we resolve the linked thread.
